_drafts/unit-testing-basics.md

```diff
@@ -0,0 +1,1257 @@
+
+{:.no_toc}
+
+### Overview
+
+We are tasked with writing a Key/Value storage module in a firmware project. The
+requirements are as follows:
+
+- The module should have the ability to read, write, and delete key/value pairs.
+- The backing data should be stored in `littlefs`.
+- The number of times a key/value pair is read, written, or deleted is counted
+  using an `analytics.c` module with a function call to `analytics_increment`.
+  This might be used to track roughly how often the flash chip is written to.
+- The module should be locked by a mutex so that only one consumer can be
+  writing, reading, or deleting from the `/kv` directory of `littlefs`.
+
+In an ideal world, and in our realistic one as well, it is possible for us to
+write this entire module _and_ test it without actually using real hardware.
+
+Let's get started!
+
+{:.no_toc}
+
+### Basic Implementation of K/V Store
+
+Below is our first attempt at `kv_store.c` which is the skeleton of our file.
+
+```c
+#include "kv_store.h"
+#include <inttypes.h>
+#include "lfs.h"
+
+void kv_store_init(lfs_t *lfs) {
+}
+
+bool kv_store_write(const char *key, const void *val, uint32_t len) {
+  return true;
+}
+
+bool kv_store_read(const char *key, void *buf,
+                   uint32_t buf_len, uint32_t *len_read) {
+  return true;
+}
+
+bool kv_store_delete(const char *key) {
+  return true;
+}
+```
+
+Believe it or not, we are ready to create a unit test to test that things are
+working. It's usually easier to write a unit test earlier rather than later
+since the number of dependencies can grow out of hand quickly.
+
+Below is a unit test which we mainly create to test compilation and the harness.
+
+```c++
+#include "CppUTest/TestHarness.h"
+
+extern "C" {
+  #include "kv_store/kv_store.h"
+}
+
+TEST_GROUP(TestKvStore) {
+  void setup() {
+  }
+
+  void teardown() {
+  }
+};
+
+TEST(TestKvStore, Test_SimpleKvStore) {
+  // Just make sure that our file is hooked up
+  LONGS_EQUAL(true, kv_store_read(NULL, NULL, 0, NULL));
+}
+```
+
+Let's run the test and see what prints!
+
+```
+$ cd complex/tests
+$ make
+compiling kv_store.c
+Building archive build/kv_store/lib/libkv_store.a
+Linking build/kv_store/kv_store_tests
+Running build/kv_store/kv_store_tests
+.
+OK (1 tests, 1 ran, 1 checks, 0 ignored, 0 filtered out, 1 ms)
+```
+
+Looks like our test passes and we are ready to move onto a (more) realistic
+test.
+
+{:.no_toc}
+
+### Add littlefs Implementation
+
+Our requirement was that our `kv_store` implementation must use `littlefs` to
+store its data. At first, the task seems daunting! How are we supposed to write
+to a filesystem that doesn't exist on our host machine? Also, a filesystem is a
+complicated piece of software!
+
+Thankfully, `littlefs` includes an _emulated_ version of its filesystem which
+runs directly on a PC. These source files are under `littlefs/emubd`, and we can
+add them to our unit test to make a fully functional `littlefs` filesystem.
```

Comment:
Maybe note that this is the "fake" for our test?

Reply:
Good idea.